File: hu.bme.mit.incqueryd.runtime/hu.bme.mit.incqueryd.engine/src/test/java/hu/bme/mit/incqueryd/engine/rete/actors/tests/ReteActorTest.java
/*******************************************************************************
 * Copyright (c) 2010-2014, Gabor Szarnyas, Istvan Rath and Daniel Varro
 * All rights reserved. This program and the accompanying materials
 * are made available under the terms of the Eclipse Public License v1.0
 * which accompanies this distribution, and is available at
 * http://www.eclipse.org/legal/epl-v10.html
 *
 * Contributors:
 * Gabor Szarnyas - initial API and implementation
 *******************************************************************************/
package hu.bme.mit.incqueryd.engine.rete.actors.tests;

import hu.bme.mit.incqueryd.engine.rete.actors.ReteActor;
import hu.bme.mit.incqueryd.engine.rete.messages.ActorReply;
import hu.bme.mit.incqueryd.engine.rete.messages.SubscriptionMessage;

import org.junit.AfterClass;
import org.junit.BeforeClass;
import org.junit.Test;

import akka.actor.ActorRef;
import akka.actor.ActorSystem;
import akka.actor.Props;
import akka.testkit.JavaTestKit;

public class ReteActorTest {

	static ActorSystem system;

	@BeforeClass
	public static void setup() {
		system = ActorSystem.create();
	}

	@AfterClass
	public static void teardown() {
		system.shutdown();
	}

	@Test
	public void subscription() {
		new JavaTestKit(system) {
			{
				// Arrange
				final Props props = new Props(ReteActor.class);
				final ActorRef actor = system.actorOf(props);

				// create probes to check the propagated
				final JavaTestKit probe1 = new JavaTestKit(system);
				final JavaTestKit probe2 = new JavaTestKit(system);

				// Act
				actor.tell(SubscriptionMessage.SUBSCRIBE_SINGLE, probe1.getRef());
				// Assert
				probe1.expectMsgEquals(duration("1 second"), ActorReply.SUBSCRIBED);

				// Act
				actor.tell(SubscriptionMessage.SUBSCRIBE_SINGLE, probe2.getRef());
				// Assert
				probe2.expectMsgEquals(duration("1 second"), ActorReply.SUBSCRIBED);				
			}
		};
	}

}

File: hu.bme.mit.incqueryd.client/hu.bme.mit.incqueryd.rete.actors/src/test/scala/hu/bme/mit/incqueryd/engine/rete/actors/tests/ReteActorTest.scala
package hu.bme.mit.incqueryd.engine.rete.actors.tests

import org.junit.AfterClass
import org.junit.BeforeClass
import org.junit.Test
import akka.actor.ActorSystem
import akka.actor.Props
import akka.testkit.JavaTestKit
import akka.testkit.JavaTestKit.duration
import hu.bme.mit.incqueryd.engine.rete.actors.RegisterSubscriber
import hu.bme.mit.incqueryd.engine.rete.actors.ReteActor
import hu.bme.mit.incqueryd.engine.rete.actors.SubscriberRegistered
import hu.bme.mit.incqueryd.engine.rete.dataunits.ReteNodeSlot
import ReteActorTest._
import eu.mondo.utils.NetworkUtils
import hu.bme.mit.incqueryd.actorservice.YarnActorService
import hu.bme.mit.incqueryd.actorservice.AkkaUtils
import org.junit.Before
import org.junit.After
import scala.concurrent.Await

object ReteActorTest {

  var system: ActorSystem = _

  @BeforeClass
  def setup() {
    system = AkkaUtils.getRemotingActorSystem(YarnActorService.actorSystemName, NetworkUtils.getLocalIpAddress, YarnActorService.port)
  }

  @AfterClass
  def teardown() {
     //Await.result(system.terminate(), AkkaUtils.defaultTimeout)
  }
}

class ReteActorTest {

  @Test
  def subscription() {
    new JavaTestKit(system) {

      val props = Props.create(classOf[ReteActor])

      val actor = system.actorOf(props)

      val probe1 = new JavaTestKit(system)

      val probe2 = new JavaTestKit(system)

      actor.tell(RegisterSubscriber(ReteNodeSlot.SINGLE), probe1.getRef)

      probe1.expectMsgEquals(duration("1 second"), SubscriberRegistered)

      actor.tell(RegisterSubscriber(ReteNodeSlot.SINGLE), probe2.getRef)

      probe2.expectMsgEquals(duration("1 second"), SubscriberRegistered)
    }
  }
}
